src/integrations/oracles/broadcaster/BroadcasterOracle.sol

```diff
@@ -0,0 +1,124 @@
+     */
+    function receiver() public view returns (IReceiver) {
+        return _receiver;
+    }
+
+    /**
+     * @notice Returns the broadcaster contract.
+     * @return The broadcaster contract.
+     */
+    function broadcaster() public view returns (IBroadcaster) {
+        return _broadcaster;
+    }
+
+    /**
+     * @notice Verifies a message broadcasted in a remote chain.
+     * @param broadcasterReadArgs The arguments required to read the broadcaster's account on the remote chain.
+     * @param remoteChainId The chain id of the remote chain.
+     * @param remoteOracle The address of the remote oracle.
+     * @param messageData The data of the message to verify.
+     */
+    function verifyMessage(
+        IReceiver.RemoteReadArgs calldata broadcasterReadArgs,
+        uint256 remoteChainId,
+        address remoteOracle, // publisher
+        bytes calldata messageData
+    ) external {
+        (bytes32 application, bytes32[] memory payloadHashes) =
+            MessageEncodingLib.getHashesOfEncodedPayloads(messageData);
+
+        bytes32 message = keccak256(abi.encodePacked(payloadHashes));
+
+        bytes32 broadcasterId = bytes32(reverseChainIdMap[remoteChainId]);
+
+        (bytes32 actualBroadcasterId,) = receiver().verifyBroadcastMessage(broadcasterReadArgs, message, remoteOracle);
+
+        if (actualBroadcasterId != broadcasterId) revert InvalidBroadcasterId();
+
+        uint256 numPayloads = payloadHashes.length;
+        for (uint256 i = 0; i < numPayloads; i++) {
+            bytes32 payloadHash = payloadHashes[i];
+            _attestations[remoteChainId][remoteOracle.toIdentifier()][application][payloadHash] = true;
+
+            emit OutputProven(remoteChainId, remoteOracle.toIdentifier(), application, payloadHash);
+        }
+    }
+
+    /**
+     * @notice Submits a proof of filled payloads as a message to the broadcaster.
+     * @param source The address of the application that has attested the payloads.
+     * @param payloads The payloads to submit.
+     */
+    function submit(
+        address source,
+        bytes[] calldata payloads
+    ) public {
+        if (payloads.length == 0) revert EmptyPayloads();
+        if (!IAttester(source).hasAttested(payloads)) revert NotAllPayloadsValid();
+
+        bytes32 message = _getMessage(payloads);
+
+        broadcaster().broadcastMessage(message);
+    }
+
+    /**
+     * @notice Generates a message from the payloads.
+     * @param payloads The payloads to generate the message from.
+     * @return message The message generated from the payloads.
+     */
+    function _getMessage(
+        bytes[] calldata payloads
+    ) internal pure returns (bytes32 message) {
+        bytes32[] memory payloadHashes = new bytes32[](payloads.length);
+        for (uint256 i = 0; i < payloads.length; i++) {
+            payloadHashes[i] = keccak256(payloads[i]);
+        }
+        return keccak256(abi.encodePacked(payloadHashes));
```

Comment:
`abi.encodePacked` should be unnecessary here, the payload hashes are in memory in the same way that we need to hash them. Can we do this using assembly?